core/src/main/scala/akka/kafka/ConsumerSettings.scala

```diff
@@ -20,21 +22,51 @@ import scala.collection.JavaConverters._
 import scala.compat.java8.OptionConverters._
 import scala.concurrent.duration._
 
-sealed trait Subscription
-sealed trait ManualSubscription extends Subscription
-sealed trait AutoSubscription extends Subscription
+sealed trait Subscription {
+  def rebalanceListener: Option[ActorRef]
+  /** Configure this actor ref to receive [[akka.kafka.ConsumerRebalanceEvent]] signals */
+  def withRebalanceListener(ref: ActorRef): Subscription
+}
+sealed trait ManualSubscription extends Subscription {
+  def withRebalanceListener(ref: ActorRef): ManualSubscription
+}
+sealed trait AutoSubscription extends Subscription {
+  def withRebalanceListener(ref: ActorRef): AutoSubscription
+}
+
+sealed trait ConsumerRebalanceEvent
+final case class TopicPartitionsAssigned(sub: Subscription, topicPartitions: Set[TopicPartition]) extends ConsumerRebalanceEvent
+final case class TopicPartitionsRevoked(sub: Subscription, topicPartitions: Set[TopicPartition]) extends ConsumerRebalanceEvent
 
 object Subscriptions {
-  private[kafka] final case class TopicSubscription(tps: Set[String]) extends AutoSubscription
-  private[kafka] final case class TopicSubscriptionPattern(pattern: String) extends AutoSubscription
-  private[kafka] final case class Assignment(tps: Set[TopicPartition]) extends ManualSubscription
-  private[kafka] final case class AssignmentWithOffset(tps: Map[TopicPartition, Long]) extends ManualSubscription
-  private[kafka] final case class AssignmentOffsetsForTimes(timestampsToSearch: Map[TopicPartition, Long]) extends ManualSubscription
+  /** INTERNAL API */
+  private[kafka] final case class TopicSubscription(tps: Set[String])(val rebalanceListener: Option[ActorRef]) extends AutoSubscription {
```

Comment:
why second parameter list?

Reply:
wanted to avoid including it in the pattern match but maybe does not matter since internal APIs anyway